src/Lean/Environment.lean

```diff
@@ -106,24 +101,172 @@ structure EnvironmentHeader where
+  kernel.
+  -/
+  private mk ::
+  /--
+  Mapping from constant name to `ConstantInfo`. It contains all constants (definitions, theorems,
+  axioms, etc) that have been already type checked by the kernel.
+  -/
+  constants   : ConstMap
+  /--
+  `quotInit = true` if the command `init_quot` has already been executed for the environment, and
+  `Quot` declarations have been added to the environment. When the flag is set, the type checker can
+  assume that the `Quot` declarations in the environment have indeed been added by the kernel and
+  not by the user.
+  -/
+  quotInit    : Bool := false
+  /--
+  Diagnostic information collected during kernel execution.
+
+  Remark: We store kernel diagnostic information in an environment field to simplify the interface
+  with the kernel implemented in C/C++. Thus, we can only track declarations in methods, such as
+  `addDecl`, which return a new environment. `Kernel.isDefEq` and `Kernel.whnf` do not update the
+  statistics. We claim this is ok since these methods are mainly used for debugging.
+  -/
+  diagnostics : Diagnostics := {}
+deriving Nonempty
+
+/-- Exceptions that can be raised by the kernel when type checking new declarations. -/
+inductive Exception where
+  | unknownConstant  (env : Environment) (name : Name)
+  | alreadyDeclared  (env : Environment) (name : Name)
+  | declTypeMismatch (env : Environment) (decl : Declaration) (givenType : Expr)
+  | declHasMVars     (env : Environment) (name : Name) (expr : Expr)
+  | declHasFVars     (env : Environment) (name : Name) (expr : Expr)
+  | funExpected      (env : Environment) (lctx : LocalContext) (expr : Expr)
+  | typeExpected     (env : Environment) (lctx : LocalContext) (expr : Expr)
+  | letTypeMismatch  (env : Environment) (lctx : LocalContext) (name : Name) (givenType : Expr) (expectedType : Expr)
+  | exprTypeMismatch (env : Environment) (lctx : LocalContext) (expr : Expr) (expectedType : Expr)
+  | appTypeMismatch  (env : Environment) (lctx : LocalContext) (app : Expr) (funType : Expr) (argType : Expr)
+  | invalidProj      (env : Environment) (lctx : LocalContext) (proj : Expr)
+  | thmTypeIsNotProp (env : Environment) (name : Name) (type : Expr)
+  | other            (msg : String)
+  | deterministicTimeout
+  | excessiveMemory
+  | deepRecursion
+  | interrupted
+
+namespace Environment
+
+@[export lean_environment_find]
+def find? (env : Environment) (n : Name) : Option ConstantInfo :=
+  /- It is safe to use `find'` because we never overwrite imported declarations. -/
+  env.constants.find?' n
+
+@[export lean_environment_mark_quot_init]
+private def markQuotInit (env : Environment) : Environment :=
+  { env with quotInit := true }
+
+@[export lean_environment_quot_init]
+private def isQuotInit (env : Environment) : Bool :=
+  env.quotInit
+
+/--
+Type check given declaration and add it to the environment
+
+**NOTE**: This function does not implement `reduceBool`/`reduceNat` special reduction rules.
```

Comment:
The lack of support for `reduceBool` / `reduceNat` in `Kernel.Environment` is creating too many issues and hacks. This API feels like a footgun.